### PR TITLE
stage2: rework AIR memory layout

### PR DESCRIPTION
**Pull request by andrewrk**
This follows in the footsteps of #7920 and #8266. It is a reboot of #9068. I [live streamed development on the beginning part of this branch](https://vimeo.com/572815502). This is the last memory layout rework planned in the series.

## Merge Checklist

 * [x] Update Air.zig
 * [x] Encoding for Assembly instruction
 * [x] Update liveness analysis
 * [x] Re-implement dumping AIR as text (for debugging/compiler development purposes)
 * [x] Rework Sema to use the new memory layout
 * [x] Rework codegen.zig to use the new memory layout
 * [x] Rework C backend to use the new memory layout
 * [x] Rework SPIR-V backend to use the new memory layout
 * [x] Rework Wasm backend to use the new memory layout
 * [x] Rework LLVM backend to use the new memory layout
 * [x] Make sure debug info for parameters is working
 * [x] switch expressions
 * [x] Passing test suite

### CODE REVIEW

**Comment by andrewrk** [comment]
### benchmark: print1mil.zig (not representative!)
comparing this branch against master 
 * :x: wall clock time: 0.53s => 0.62 (17% worse)
 * :heavy_check_mark: peak memory usage: 357 MiB => 290 MiB (19% improvement)
 * :heavy_check_mark: cache-misses: 18% improvement
 * :heavy_check_mark: cache-references: 4% improvement
 * :x: instructions: 12% worse
 * :x: branches: 7% worse

**Comment by andrewrk** [comment]
Benchmark update after that last commit:

### benchmark: print1mil.zig (not representative!)

comparing this branch against master 

 * :heavy_check_mark: wall clock time: 0.54s => 0.42s (22% faster)
 * :heavy_check_mark: peak memory usage: 358 MiB => 187 MiB (48% less memory)
 * :heavy_check_mark: cache-misses: 48% fewer
 * :heavy_check_mark: cache-references: 21% fewer
 * :heavy_check_mark: instructions: 9% fewer
 * :heavy_check_mark: branches: 13% fewer

This (very contrived benchmark) is 2.4 million lines per second on my laptop.

**Comment by LewisGaul** [comment]
The CBE tests are hitting the following added `unreachable` in [`Compilation.performAllTheWork()`](https://github.com/ziglang/zig/pull/9353/commits/bed34fd03c5c3903d87fab92cb6e00e754ee1884?file-filters%5B%5D=.zig#diff-a95a8b6c400e686284af9c4312de54b1ff51ec93d2d54f483f7ba08b053ebc70R2021):
```diff
-                        .success => {},
+                        .success => unreachable, // don't queue it twice
```

If this is changed from `unreachable` to `continue` then 11/22 of the CBE tests pass for me (with g-w1's patch applied).

The first testcase that hits this can be simplified to:
```zig
var case = ctx.exeFromCompiledC("@intToError", .{});
case.addCompareOutput(
    \\pub export fn main() c_int {
    \\    // runtime checks
    \\    var x = error.A;
    \\    var y = error.B;
    \\    var z = @intToError(2);
    \\    var f = @intToError(1);
    \\    if (!(y == z)) unreachable;
    \\    if (!(x == f)) unreachable;
    \\    return 0;
    \\}
, "");
```

I've done some stepping through in GDB but my understanding of the flows is limited at this point. I thought I might as well report my findings at this point in case they're helpful.

File: src/Sema.zig [hunk 1]
```diff
@@ -2636,10 +2634,12 @@ fn zirMergeErrorSets(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) Com
         };
         return sema.mod.failWithOwnedErrorMsg(&block.base, msg);
     }
-    if (rhs_ty.zigTypeTag() != .ErrorSet)
-        return sema.mod.fail(&block.base, rhs_src, "expected error set type, found {}", .{rhs_ty});
+    const lhs_ty = try sema.analyzeAsType(block, lhs_src, lhs);
+    const rhs_ty = try sema.analyzeAsType(block, rhs_src, rhs);
     if (lhs_ty.zigTypeTag() != .ErrorSet)
         return sema.mod.fail(&block.base, lhs_src, "expected error set type, found {}", .{lhs_ty});
+    if (rhs_ty.zigTypeTag() != .ErrorSet)
```
Comment (miscco) [inline src/Sema.zig:2641]: Wont that eat a diagnostic if both are erroneous?

Comment (andrewrk) [reply src/Sema.zig:2641]: should be fine. if you provide a test case, I'm happy to look at this idea in more detail. Either way, for the purposes of this branch, all we need to do is match master branch.

**Comment by frmdstryr** [comment]
Can a breaking label be added to this PR?  

Requiring string literals for assembly sort of defeats the usefulness of comptime...

**Comment by andrewrk** [comment]
> Requiring string literals for assembly sort of defeats the usefulness of comptime...

@N00byEdge and I had a discussion about this topic and came to the conclusion that although the plans for inline assembly in zig ultimately will be not based on string concatenation, in the near term string concatenation is making up for missing features. So I will revert this requirement soon. However, eventually, the plan is to integrate inline assembly more closely with the language, in such a way that text-based substitution and string concatenation is not possible. Give it a couple weeks then check back on inline assembly in master branch.

**Comment by frmdstryr** [comment]
Only used it in a few places so its not too big of a deal..

> However, eventually, the plan is to integrate inline assembly more closely with the language, in such a way that text-based substitution and string concatenation is not even desirable.

^^ this would be sweet

**Review by geon — COMMENTED** [review]
Creating the todo should have been a separate commit. As it is, git is confused, and the diff is unreadable.

**Pull request merged**